mgr/rbd_support: recover from rados client blocklisting

ajarr opened the pull request:
In certain scenarios the OSDs were slow to process RBD requests.
This lead to the rbd_support module's RBD client not being able to
gracefully handover a RBD exclusive lock to another RBD client.
After the condition persisted for some time, the other RBD client
forcefully acquired the lock by blocklisting the rbd_support module's
RBD client, and consequently blocklisted the module's RADOS client. The
rbd_support module stopped working. To recover the module, the entire
mgr service had to be restarted which reloaded other mgr modules.

Instead of recovering the rbd_support module from client blocklisting
by being disruptive to other mgr modules, recover the module
automatically without restarting the mgr serivce. On client getting
blocklisted, shutdown the module's handlers and blocklisted client,
create a new rados client for the module, and start the new handlers.

Fixes: https://tracker.ceph.com/issues/56724
Signed-off-by: Ramana Raja <rraja@redhat.com>
<!--
  - Please give your pull request a title like

      [component]: [short description]

  - Please use this format for each git commit message:

      [component]: [short description]

      [A longer multiline description]

      Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
      Signed-off-by: [Your Name] <[your email]>

    For examples, use "git log".
-->

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [x] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

idryomov commented on src/pybind/mgr/rbd_support/mirror_snapshot_schedule.py:
```diff
@@ -350,6 +370,12 @@ def run(self) -> None:
             self.log.info("MirrorSnapshotScheduleHandler: starting")
             while True:
                 refresh_delay = self.refresh_images()
+                if self.need_to_reconnect:
```
Is `need_to_reconnect` flag actually needed?  Can `BlocklistedError` exception just be propagated and handled here?  A bit less state to manage that way.

ajarr replied:
good point. Done.

ajarr commented on src/pybind/mgr/rbd_support/mirror_snapshot_schedule.py:
```diff
@@ -335,13 +335,33 @@ def __init__(self, module: Any) -> None:
+        if self.rados:
+            self.rados.shutdown()
+
     def _cleanup(self) -> None:
```
_cleanup() is not used. removed it.

ajarr commented on src/pybind/mgr/rbd_support/mirror_snapshot_schedule.py:
```diff
@@ -288,6 +287,7 @@ def finish(self, image_spec: ImageSpec) -> None:
 
         with self.lock:
             self.pending.remove(image_spec)
+            self.condition.notify()
```
removed _cleanup() that is never called.

ajarr commented on src/pybind/mgr/rbd_support/mirror_snapshot_schedule.py:
```diff
@@ -335,13 +335,33 @@ def __init__(self, module: Any) -> None:
+        self.rados.connect()
+        self.log.info("MirrorSnapshotScheduleHandler: connected own RADOS client")
+
+    def shutdown(self) -> None:
```
makes sense.  In the latest patchset, `wait_for_pending()` is called in `shutdown()`.

ajarr commented on qa/workunits/rbd/cli_generic.sh:
```diff
@@ -1358,6 +1358,52 @@ test_mirror_snapshot_schedule() {
     ceph osd pool rm rbd2 rbd2 --yes-i-really-really-mean-it
 }
 
+test_mirror_snapshot_schedule_recovery() {
```
Couple of notes regarding this test:
- Maybe we want to combine this with test_mirror_snapshot_schedule? But I didn't do that for now considering that https://tracker.ceph.com/issues/58490 causes the test_mirror_snapshot_schedule to fail

- This test uncovered a bug in the changes I made to  `rbd_support.schedule.LevelSpec.from_name`, where I replaced `module.rados` with `handler.rados` to allow the scheduler handlers to only use handler's RADOS client.
I hit the following error,
```
2023-02-04T20:22:17.389-0500 7f9035f8a640  0 [rbd_support ERROR root] Fatal runtime error:
Traceback (most recent call last):
File "/home/rraja/git/ceph/src/pybind/mgr/rbd_support/module.py", line 42, in wrapper
    return func(self, *args, **kwargs)
File "/home/rraja/git/ceph/src/pybind/mgr/rbd_support/module.py", line 102, in mirror_snapshot_schedule_add
    spec = LevelSpec.from_name(
 File "/home/rraja/git/ceph/src/pybind/mgr/rbd_support/schedule.py", line 125, in from_name
    raise ValueError("{} is not an RBD pool".format(pool_name))
ValueError: rbd3 is not an RBD pool
```
This was because the pool ID of `rbd3` fetched using `handler.rados.pool_lookup(pool_name)` (pool ID was 8) was different from the pool ID of `rbd3` in the output of `get_rbd_pools(handler.module)` (pool ID was 9) .  I'm not sure why this happens? So in my local testing I reverted the changes I made to `rbd_support.schedule.LevelSpec.from_name` to continue using  `module.rados`, which fixed this issue.

trociny replied:
I suppose it is because now your (not module's) rados client does not wait for the latest osd map, and you use the ID (8) of rbd3 pool that was created (and deleted) for the previous test.

Please look at the `with_latest_osdmap` decorator in `module.py` and how it is used to make sure all commands that need the latest osdmap wait for it. You need this for your rados instances too.

idryomov replied:
> This was because the pool ID of `rbd3` fetched using `handler.rados.pool_lookup(pool_name)` (pool ID was 8) was different from the pool ID of `rbd3` in the output of `get_rbd_pools(handler.module)` (pool ID was 9) .

Is this reproducible?  Always or only if `test_trash_purge_schedule_recovery` (or anything else) is executed before `test_mirror_snapshot_schedule_recovery`?  I suspect the root cause boils down to the fact that different RADOS instances may have (be aware of) different versions of osdmap.  It's not uncommon for a specific RADOS instances to be a little "behind" and you are re-creating a pool with same name here.  Also notice that:
- `get_rbd_pools` operates on the osdmap that is returned directly by the manager framework (it's cached)
- `with_latest_osdmap` in `src/pybind/mgr/rbd_support/module.py` syncs up with the latest osdmap using the module-wide RADOS instance

idryomov replied:
Ah, it looks like @trociny beat me to it -- my comments sat there waiting for "Finish review" to be clicked ;)

ajarr replied:
> I suppose it is because now your (not module's) rados client does not wait for the latest osd map, and you use the ID (8) of rbd3 pool that was created (and deleted) for the previous test.
> 
> Please look at the `with_latest_osdmap` decorator in `module.py` and how it is used to make sure all commands that need the latest osdmap wait for it. You need this for your rados instances too.

Thanks @trociny ! I missed making the rados instances wait for latest osdmap when connecting and when a command was rbd_support module command handled. I did that and the tests passed.

ajarr replied:
> > This was because the pool ID of `rbd3` fetched using `handler.rados.pool_lookup(pool_name)` (pool ID was 8) was different from the pool ID of `rbd3` in the output of `get_rbd_pools(handler.module)` (pool ID was 9) .
> 
> Is this reproducible? Always or only if `test_trash_purge_schedule_recovery` (or anything else) is executed before `test_mirror_snapshot_schedule_recovery`? 

I didn't hit this bug in my local testing. I assumed  I'll hit it when running this PR in teuthology.

> I suspect the root cause boils down to the fact that different RADOS instances may have (be aware of) different versions of osdmap. It's not uncommon for a specific RADOS instances to be a little "behind" and you are re-creating a pool with same name here. Also notice that:
> 
> * `get_rbd_pools` operates on the osdmap that is returned directly by the manager framework (it's cached)
> * `with_latest_osdmap` in `src/pybind/mgr/rbd_support/module.py` syncs up with the latest osdmap using the module-wide RADOS instance

Thanks for the explanation. I dug further and saw that `get_rbd_pools` eventually calls into `pymodules->cacheable_get_python` . Yes, I missed making the RADOS instances of the handlers wait for the latest osdmap.

idryomov commented on src/mgr/PyModuleRegistry.h:
```diff
@@ -227,5 +227,15 @@ class PyModuleRegistry
     return v;
   }
 
+  auto get_clients_names() const
```
... and this method could be changed to return `clients` multimap directly.

idryomov commented on src/pybind/rados/rados.pyx:
```diff
@@ -224,6 +224,13 @@ class IsConnected(Error):
                 "RADOS is connected error (%s)" % message, errno)
 
 
+class BlocklistedError(Error):
```
I'd rename to `ConnectionShutdown` to match the one in the RBD bindings (and the actual errno).

ajarr replied:
Done.

idryomov reviewed (commented):
I focused on the mgrmap bits since they are going to be useful independent of what we end up doing for the `rbd_support` module.

ajarr commented:
Based on the RBD standup discussion today (Feb 7, 2023), the plan is to split up this PR. The work related to updating the mgrmap to store active_client's module name will be uploaded as a separate PR.

The work related to adding separate RADOS clients for the {mirror_snapshot_schedule, trash_purge_schedule, task} handlers to allow recovery of the rbd_support module in environment where the module's RADOS instances get blocklisted,  is also being rethought. If it's possible to have the rbd_support module's RADOS client reconnect on being blocklisted, then maybe that's enough?

ajarr commented on src/messages/MMgrBeacon.h:
```diff
@@ -47,6 +47,8 @@ class MMgrBeacon final : public PaxosServiceMessage {
 
   std::vector<entity_addrvec_t> clients;
 
+  std::vector<std::string> clients_names;
```
If we're decoding a version of the message that doesn't have the vector of client_names, how do we the populate the `clients` object with datatype `std::multimap<std::string, entity_addrvec_t>`?  Are you suggesting that we have two data members in the class MMgrBeacon, one `std::vector<entity_addrvec_t> clients` and the other `std::multimap<std::string, entity_addrvec_t> clients_with_names`?   And similarly make change to the MgrMap?

idryomov replied:
One `std::multimap` member named `clients`, use an empty string as the name if the message doesn't have names.  Multimap permits multiple values with the same key (that's what it's for) so you would just end up with all addresses under that single `""` key.

ajarr replied:
Thanks! Addressed your comments related to this commit in https://github.com/ceph/ceph/pull/50065

ajarr commented:
> Based on the RBD standup discussion today (Feb 7, 2023), the plan is to split up this PR. The work related to updating the mgrmap to store active_client's module name will be uploaded as a separate PR.
> 

Separate PR https://github.com/ceph/ceph/pull/50065 that adds changes to MgrMap

ajarr commented:
Discussed with @idryomov. Recovering the module from the mirror_snapshot_schedule handler's event loop will make the handler special. This breaks the existing model where a handle is not dependant on another. Instead, he suggested creating a recovery thread in the mgr module that periodically checks a boolean flag, which is set when the module's rados client is blocklisted. When the flag is True, the recovery thread will call each of the handler's shutdown method that'd cause the handler's thread to exit (and in the case of mirror_snapshot_schedule handler, also wait for the queue of async pending snapshots to empty). Once all the handler shutdowns complete, the module's rados client will be shutdown, a new connection will be started, and a new set of handlers will be created.

ajarr commented:
@idryomov I've updated the draft as per suggestions. PTAL

ajarr commented on src/pybind/mgr/rbd_support/task.py:
```diff
@@ -170,6 +170,7 @@ def __init__(self, module: Any) -> None:
         with self.lock:
             self.init_task_queue()
 
+        self.stop_thread = True
```
self.stop_thread = False

ajarr commented:
In the latest patchset,
* added tests for recovery of perf and task handlers
* made minor changes to tests for recovery of mirror_snapshot_schedule and trash_purge_schedule handlers
* fixed an issue in task handler identified during testing. See, https://github.com/ceph/ceph/pull/49742#discussion_r1141465965

idryomov reviewed (commented):
This approach seems right!  I left a few nits, some (`except` clauses and debug message uniformity) apply throughout the PR.  Please squash all `cli_generic.sh` commits into a single commit, I'll take a closer look at it next time.

ajarr commented on src/pybind/mgr/rbd_support/module.py:
```diff
@@ -74,11 +83,50 @@ class Module(MgrModule):
+        self.client_blocklisted = Event()
+        self.setup()
+        self.recovery_thread = Thread(target=self.run)
+        self.recovery_thread.start()
```
Good point!

ajarr commented on src/pybind/mgr/rbd_support/module.py:
```diff
@@ -46,6 +51,10 @@ def wrapper(self: 'Module', *args: Any, **kwargs: Any) -> Tuple[int, str, str]:
                 self.log.exception("Fatal runtime error: ")
                 raise
+        except (rados.ConnectionShutdown, rbd.ConnectionShutdown) as ex:
+            self.log.debug("with_latest_osdmap: client is blocklisted")
```
Done

ajarr commented on src/pybind/mgr/rbd_support/module.py:
```diff
@@ -74,11 +83,50 @@ class Module(MgrModule):
+                # block until rados client is blocklisted
+                self.client_blocklisted.wait()
+                self.restart()
+                sleep(10)
```
It's not required

ajarr commented on src/pybind/rados/rados.pyx:
```diff
@@ -224,6 +224,13 @@ class IsConnected(Error):
                 "RADOS is connected error (%s)" % message, errno)
 
 
+class ConnectionShutdown(OSError):
```
I'm not sure too. It does seem arbitrary. Let me know if you prefer one over the other?

ajarr replied:
Thanks for figuring it out. I'll leave it as is then.

ajarr commented:
> The approach seems right to me! I left a few nits, some (`except` clauses and debug message uniformity) apply throughout the PR. Please squash all `cli_generic.sh` commits into a single commit, I'll take a closer look at it next time.

Thanks for the review, @idryomov . I think I've addressed all your requested changes, and made the changes uniformly across all the handlers.

ajarr commented:
I see mypy failures here, https://jenkins.ceph.com/job/ceph-pull-requests/113916/testReport/junit/projectroot.src.pybind/mgr/run_tox_mgr/

ajarr commented:
> I see mypy failures here, https://jenkins.ceph.com/job/ceph-pull-requests/113916/testReport/junit/projectroot.src.pybind/mgr/run_tox_mgr/

Fixed them

ajarr commented on src/pybind/mgr/rbd_support/mirror_snapshot_schedule.py:
```diff
@@ -330,7 +331,6 @@ class MirrorSnapshotScheduleHandler:
 
     lock = Lock()
     condition = Condition(lock)
-    thread = None
```
Hit the following mypy error for all the handlers due to my changes to the handlers' `shutdown()` method,
```
rbd_support/module.py:18: note: In module imported here,
rbd_support/__init__.py:2: note: ... from here:
rbd_support/mirror_snapshot_schedule.py: note: In member "shutdown" of class "MirrorSnapshotScheduleHandler":
rbd_support/mirror_snapshot_schedule.py:351: error: Item "None" of "Optional[Thread]" has no attribute "is_alive"
rbd_support/mirror_snapshot_schedule.py:353: error: Item "None" of "Optional[Thread]" has no attribute "join"
Found 6 errors in 3 files (checked 32 source files)
```
I removed the error by making `thread` an instance attribute. I can also get rid of the error by adding the following line in `shutdown()` method
```
assert self.thread is not None
```

I'm not sure why the handlers have `thread` as a class attribute?

ajarr commented:
jenkins test make check

ajarr commented:
jenkins test make check arm64

idryomov commented on src/pybind/mgr/rbd_support/task.py:
```diff
@@ -170,6 +169,7 @@ def __init__(self, module: Any) -> None:
         with self.lock:
             self.init_task_queue()
```
What happens if we get blocklisted while initializing the queue?  I wonder if this should be moved to `run()` so that potential `ConnectionShutdown` exceptions are handled.

ajarr replied:
I audited other 3 handlers. I should make similar changes to mirror_snapshot_schedule and trash_purge_schedule handlers as their `init_schedule_queue()` also use module's rados client that can get blocklisted.

idryomov replied:
Good catch!

ajarr replied:
I've made the changes. Let me know if you think it's look okay.

idryomov replied:
> I've made the changes. Let me know if you think it's look okay.

Sorry, I think my suggestion to move the entire `init_XYZ()` method to `run()` was stupid.  This means that we would set `self.module_ready` and, even more importantly, return from `Module.__init__()` without handlers actually being initialized in the sense that mirror snapshot scheduler for example wouldn't have `self.queue` attribute!

Let's revert these changes for now and think about whether anything extra is needed at all.  The failure mode is not that bad: if any exception is thrown during module initialization, it would lead to `HEALTH_ERR` status due to an always-on module failing to load.  If it happens to be due to blocklisting, then yes, we don't recover, but it's _very_ noticeable.

ajarr replied:
Done

idryomov commented:
Overall, I think there is room for improvement in the tests:
- for `TrashPurgeScheduleHandler` and `MirrorSnapshotScheduleHandler` tests, instead of just checking that both pre- and post-blocklisting schedules show up, it would be good to test that both pre- and post-blocklisting scheduled work actually gets done.  For trash purge scheduler, I would suggest creating two pools with an image trashed in each, adding a short (1-2m) schedule for one pool before blocklisting and for another after blocklisting and asserting that trash purge runs after enough time passes.  And similarly for mirror snapshot scheduler: a single pool with two images would do as per-image schedules can be added instead of per-pool there.
- for `TaskHandler` test, queue some (let's say 5) flattens on different images before blocklisting and a sixth flatten after blocklisting and assert that all six flattens complete.

ajarr commented:
@idryomov accidentally requested for re-review. pls ignore

idryomov commented:
Current tests appear to be stable with the following fixups:
```
diff --git a/qa/workunits/rbd/cli_generic.sh b/qa/workunits/rbd/cli_generic.sh
index 46160c8059df..57279d26dcee 100755
--- a/qa/workunits/rbd/cli_generic.sh
+++ b/qa/workunits/rbd/cli_generic.sh
@@ -1503,13 +1503,15 @@ test_perf_image_iostat_recovery() {
     echo "testing recovery of perf handler after module's RADOS client is blocklisted..."
     remove_images
 
-    ceph osd pool create rbd1 8
-    rbd pool init rbd1
-    rbd namespace create rbd1/ns
+    ceph osd pool create rbd3 8
+    rbd pool init rbd3
+    rbd namespace create rbd3/ns
 
-    IMAGE_SPECS=("rbd1/test1" "rbd1/ns/test2")
+    IMAGE_SPECS=("rbd3/test1" "rbd3/ns/test2")
     for spec in "${IMAGE_SPECS[@]}"; do
-        rbd create $RBD_CREATE_ARGS --size 10G $spec
+        # ensure all images are created without a separate data pool
+        # as we filter iostat by specific pool specs below
+        rbd create $RBD_CREATE_ARGS --size 10G --rbd-default-data-pool '' $spec
     done
 
     BENCH_PIDS=()
@@ -1519,7 +1521,7 @@ test_perf_image_iostat_recovery() {
         BENCH_PIDS+=($!)
     done
 
-    test "$(rbd perf image iostat --format json rbd1 |
+    test "$(rbd perf image iostat --format json rbd3 |
         jq -r 'map(.image) | sort | join(" ")')" = 'test1'
 
     # Fetch and blocklist the rbd_support module's RADOS client
@@ -1529,10 +1531,10 @@ test_perf_image_iostat_recovery() {
     ceph osd blocklist add $CLIENT_ADDR
     ceph osd blocklist ls | grep $CLIENT_ADDR
 
-    expect_fail rbd perf image iostat --format json rbd1/ns
+    expect_fail rbd perf image iostat --format json rbd3/ns
     sleep 10
     for i in `seq 24`; do
-        test "$(rbd perf image iostat --format json rbd1/ns |
+        test "$(rbd perf image iostat --format json rbd3/ns |
             jq -r 'map(.image) | sort | join(" ")')" = 'test2' && break
 	sleep 10
     done
@@ -1543,7 +1545,7 @@ test_perf_image_iostat_recovery() {
     wait
 
     remove_images
-    ceph osd pool rm rbd1 rbd1 --yes-i-really-really-mean-it
+    ceph osd pool rm rbd3 rbd3 --yes-i-really-really-mean-it
 }
 
 test_mirror_pool_peer_bootstrap_create() {
@@ -1661,7 +1663,6 @@ test_tasks_recovery() {
     ceph osd blocklist add $CLIENT_ADDR
     ceph osd blocklist ls | grep $CLIENT_ADDR
 
-    test "$(ceph rbd task list)" = "[]"
     expect_fail ceph rbd task add flatten rbd2/clone1
     sleep 10
     for i in `seq 24`; do
@@ -1670,10 +1671,6 @@ test_tasks_recovery() {
     done
     test "$(ceph rbd task list)" != "[]"
 
-    # queue flatten and check that it completes
-    rbd info rbd2/clone1 | grep 'parent: '
-    expect_fail rbd snap unprotect rbd2/img1@snap
-    ceph rbd task add flatten rbd2/clone1
     for i in {1..12}; do
         rbd info rbd2/clone1 | grep 'parent: ' || break
         sleep 10
```

ajarr commented on src/pybind/mgr/rbd_support/mirror_snapshot_schedule.py:
```diff
@@ -450,6 +461,8 @@ def load_pool_images(self,
                     self.log.debug(
                         "load_pool_images: adding image {}".format(name))
                     images[pool_id][namespace][image_id] = name
+        except (rados.ConnectionShutdown, rbd.ConnectionShutdown):
```
On module's client being blocklisted, I wasn't sure whether `ioctx.set_namespace()`  within this try-except block would raise `rados.ConnectionShutdown()`?

ajarr commented:
> Overall, I think there is room for improvement in the tests (probably OK to defer to another PR though):
> 
> * for `TrashPurgeScheduleHandler` and `MirrorSnapshotScheduleHandler` tests, instead of just checking that both pre- and post-blocklisting schedules show up, it would be good to test that both pre- and post-blocklisting scheduled work actually gets done.  For trash purge scheduler, I would suggest creating two pools with an image trashed in each, adding a short (1-2m) schedule for one pool before blocklisting and for another after blocklisting and asserting that trash purge runs after enough time passes.  And similarly for mirror snapshot scheduler: a single pool with two images would do as per-image schedules can be added there.
> * for `TaskHandler` test, queue some (let's say 5) flattens on different images before blocklisting and a sixth flatten after blocklisting and assert that all six flattens complete.

Created tracker ticket for now, https://tracker.ceph.com/issues/59681

idryomov commented:
No related failures:

https://pulpito.ceph.com/dis-2023-05-08_22:48:48-rbd-wip-dis-testing-distro-default-smithi/
https://pulpito.ceph.com/dis-2023-05-09_11:02:54-rbd-wip-dis-testing-distro-default-smithi/
https://pulpito.ceph.com/dis-2023-05-09_20:53:52-rbd-wip-dis-testing-distro-default-smithi/

This is with https://github.com/ceph/ceph/pull/49975 excluded in the last rerun -- it's causing "Exiting scrub checking -- not all pgs scrubbed." errors.  Per @neha-ojha the plan is to introduce a more aggressive QoS profile for teuthology tests.
